src/common/io.cc

```diff
@@ -120,27 +121,30 @@ std::string LoadSequentialFile(std::string fname) {
 #if defined(__linux__)
   posix_fadvise(fd, 0, 0, POSIX_FADV_SEQUENTIAL);
 #endif  // defined(__linux__)
-  ssize_t bytes_read = read(fd, &buffer[0], f_size_bytes);
-  if (bytes_read < 0) {
-    close(fd);
-    ReadErr();
+  ssize_t bytes_read = 0;
+  while (bytes_read < f_size_bytes) {
+    ssize_t result = read(fd, &buffer[bytes_read], f_size_bytes);
+    if (result < 0) {
+      close(fd);
+      ReadErr();
+    }
+    bytes_read += result;
   }
   close(fd);
-#else  // defined(__unix__)
-  FILE *f = fopen(fname.c_str(), "r");
-  if (f == NULL) {
-    std::string msg;
-    OpenErr();
+  buffer.back() = '\0';
+  // Trim NUL letters at the end
+  int64_t i;
+  for (i = static_cast<int64_t>(buffer.size()); i >= 0; --i) {
+    if (buffer[i - 1] != '\0') {
+      break;
+    }
   }
-  fseek(f, 0, SEEK_END);
-  auto fsize = ftell(f);
-  fseek(f, 0, SEEK_SET);
-
-  buffer.resize(fsize + 1);
-  fread(&buffer[0], 1, fsize, f);
-  fclose(f);
+  buffer = buffer.substr(0, i);
```

Comment:
Why is it a substr?

Reply:
My bad. It should be resize() so that no copy is made.

Reply:
Never mind. I’ll remove this change.